src/interval_sets.jl

```diff
@@ -1,28 +1,73 @@
 
 ###### Set-related Helpers #####
 """
-    IntervalSet{T<:AbstractInterval} <: AbstractSet{T}
-
-A set type for performing multi-interval specific operations.
-https://en.wikipedia.org/wiki/Interval_arithmetic#Interval_operators
+    IntervalSet{T<:AbstractInterval}
+
+An set of points represented by a sequence of intervals. Set operations over interval sets
+return a new IntervalSet, with the fewest number of intervals possible. Unbounded intervals
+are not supported. The individual intervals in the set can be accessed using the iteration
+API or by passing the set to `Array`.
+
+see also: https://en.wikipedia.org/wiki/Interval_arithmetic#Interval_operators
+
+## Examples
+
+```jldoctest
+julia> using Intervals
+julia> convert(Array, union(IntervalSet(1..5), IntervalSet(3..8)))
+1-element Vector{Interval{Int64, Closed, Closed}}:
+ Interval{Int64, Closed, Closed}(1, 8)
+
+julia> convert(Array, intersect(IntervalSet(1..5), IntervalSet(3..8)))
+1-element Vector{Interval{Int64, Closed, Closed}}:
+ Interval{Int64, Closed, Closed}(3, 5)
+ 
+julia> convert(Array, symdiff(IntervalSet(1..5), IntervalSet(3..8)))
+2-element Vector{Interval{Int64}}:
+ Interval{Int64, Closed, Open}(1, 3)
+ Interval{Int64, Open, Closed}(5, 8)
+
+julia> convert(Array, union(IntervalSet([1..2, 2..5]), IntervalSet(6..7)))
+2-element Vector{Interval{Int64, Closed, Closed}}:
+ Interval{Int64, Closed, Closed}(1, 5)
+ Interval{Int64, Closed, Closed}(6, 7)
+
+julia> convert(Array, union(IntervalSet([1..5, 8..10]), IntervalSet([4..9, 12..14])))
+2-element Vector{Interval{Int64, Closed, Closed}}:
+ Interval{Int64, Closed, Closed}(1, 10)
+ Interval{Int64, Closed, Closed}(12, 14)
+
+julia> convert(Array, intersect(IntervalSet([1..5, 8..10]), IntervalSet([4..9, 12..14])))
+2-element Vector{Interval{Int64, Closed, Closed}}:
+ Interval{Int64, Closed, Closed}(4, 5)
+ Interval{Int64, Closed, Closed}(8, 9)
+
+julia> convert(Array, setdiff(IntervalSet([1..5, 8..10]), IntervalSet([4..9, 12..14])))
+2-element Vector{Interval{Int64}}:
+ Interval{Int64, Closed, Open}(1, 4)
+ Interval{Int64, Open, Closed}(9, 10)
+```
 """
-struct IntervalSet{T<:AbstractInterval} <: AbstractSet{T}
-    # TODO: Use Dict internally once union!(Vector{AbstractInterval}) is fully deprecated
-    items::Vector{T}
+struct IntervalSet{I <: AbstractInterval, A <: AbstractVector{I}}
+    items::A
 end
```

Comment:
> I've made it possible for `IntervalSet`'s field to be any `AbstractVector`. There are some cases where I have a view of TimeSpan.jl arrays that I want to be treated as Interval.jl arrays, I use an `AbstractArray` that converts the objects on the fly during `getindex`.

I think this is most like a premature optimization. The internal representation of an `IntervalSet` could change in the future to no longer be ordered so we can't always guarantee that this will avoid a copy.

Reply:
My vote is to define this as:
```suggestion
struct IntervalSet{T <: AbstractInterval}
    items::Vector{T}
end
```

Reply:
Fair enough. And actually the main reason I needed this no longer exists because it was mostly an issue when `IntervalSet` was called as a part of `find_intersections`, which I realized doesn't actually make sense there.